src/Sparrow/AsyncManualResetEvent.cs

```diff
@@ -8,16 +8,32 @@ namespace Sparrow
     public class AsyncManualResetEvent
     {
         private volatile TaskCompletionSource<bool> _tcs = new TaskCompletionSource<bool>();
+        private readonly CancellationToken _token;
 
-        public Task WaitAsync()
+        public AsyncManualResetEvent()
+        {
+            _token = CancellationToken.None;		    
+        }
+
+        public AsyncManualResetEvent(CancellationToken token)
+        {
+            _token = token;
+            // ReSharper disable once ImpureMethodCallOnReadonlyValueField
+            _token.Register(() => _tcs.TrySetResult(false));
```

Comment:
I wouldn't trust it this way, better to send the token to the Wait method directly